### PR TITLE
[Distributed] Cache checking distributed func

### PR DESCRIPTION
**Pull request by ktoso**
By caching distributed func checking, we're able to invoke it wherever we want: in typechecking as well as to guard the synthesis of the distributed funcs.

resolves rdar://91562751

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Comment by xedin** [comment]
@swift-ci please smoke test Linux

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

File: lib/Sema/TypeCheckDistributed.cpp [hunk 1]
```diff
@@ -434,11 +433,17 @@ static bool checkDistributedTargetResultType(
...
-/// \param diagnose Whether to emit a diagnostic when a problem is encountered.
-///
 /// \returns \c true if there was a problem with adding the attribute, \c false
 /// otherwise.
-bool swift::checkDistributedFunction(FuncDecl *func, bool diagnose) {
+bool swift::checkDistributedFunction(AbstractFunctionDecl *func) {
+  auto &C = func->getASTContext();
+  return evaluateOrDefault(C.evaluator,
+                           CheckDistributedFunctionRequest{func},
+                           true);
```
Comment (xedin) [inline lib/Sema/TypeCheckDistributed.cpp:442]: I think if there was a cycle in the graph the default should be false.

**Review by xedin — APPROVED** [review]

File: lib/Sema/TypeCheckDistributed.cpp (hunk 1 above)
Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp:442]: Okey, let's do that

Comment (ktoso) [reply lib/Sema/TypeCheckDistributed.cpp]: done

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Comment by xedin** [comment]
Thanks!

**Pull request merged**